### DISCUSSION TITLE
Rayon performance on my Gentoo Linux 6.6.30 is wierd

### DISCUSSION THREAD

**Discussion by ZhennanWu** [rayon-rs/rayon#1180] [Show and tell]
I'm not raising this as an issue but just to check if this is already known. I can't confirm whether it is rayon's issue or my Linux setup's issue for now.

# Background
[EPGI](https://github.com/ZhennanWu/epgi) is an experimental project to test the impact of parallel speed-up in GUI applications. It is designed to fully utilize rayon's ability to parallel build and layout widget.

In an extreme example app, the framework is asked to show 40k moving, interative widgets all on the same screen. This is an optimal case to test how much rayon can speed things up. And compare against other single-threaded framework.

https://github.com/ZhennanWu/epgi/assets/38578020/d08b227b-ba9b-4829-9e36-932ea48ef0db

# Setup
- Intel core i7-12700k (8 P-cores enabled, E-cores disabled). A mediocre air cooling setup.
- Windows 11
- Gentoo Linux 6.6.30 (TODO: Add relevant kernel configs)
# Results
## Success on Windows
| N_thread        | build+layout(low, ms) | paint(low, ms) | raster(ms) | FPS(high) |
|-----------------|-----------------------|----------------|------------|-----------|
|               1 |                  31.9 |            9.0 |        1.5 |      20.3 |
|               2 |                  19.8 |            9.2 |        1.4 |      28.5 |
|               4 |                  12.4 |            8.9 |        1.5 |      37.6 |
|               8 |                   8.4 |            9.1 |        1.5 |      45.1 |
|              16 |                   7.0 |            9.5 |        1.4 |      47.4 |
| Flutter Desktop |   NA due to impl diff |            ~16 |        ~28 |        16 |

The main point to checkout is `build+layout` time since currently they are the only two stages in the pipeline that are parallelized.

Things are fairly nice and stable. We have a 4x speed up using 8 threads on a 8-core machine.

## Wierd behavior on Linux
Things were less optimistic on Linux

| N               | build+layout | build+layout (low) | paint | paint (low) | raster | FPS  | FPS (high) |
|-----------------|--------------|--------------------|-------|-------------|--------|------|------------|
|               1 |         20.9 |               18.3 |   7.9 |         6.7 |    0.7 | 21.7 |       31.8 |
|               2 |         13.1 |               11.9 |   7.7 |           7 |    0.6 | 31.6 |       41.5 |
|               4 |         10.7 |                8.3 |   7.8 |         7.1 |    0.7 | 36.1 |       51.1 |
|               8 |         20.5 |               11.1 |  10.2 |         7.7 |    0.7 | 21.7 |       43.2 |
|              16 |         19.1 |               10.1 |   9.2 |         7.4 |    0.7 | 24.3 |       44.9 |
| Flutter Desktop |              |                    |       |             |        |      |         17 |

### Throttling
Unlike on Windows, the app immediately starts sharp throttling after ~5 seconds. The more thread I open up, the harsher the throttling kicks in. For each type of the metrics, "low" means the metric from best performing second, and else it means the average metric throughout a ~1min sustained execution. When using 8 threads, the app throttles from 43 fps down to 22 fps.

### Worse performance when > 4 threads
Despite my machine has 8 cores, using 8-thread and 16-thread rayon threadpool generates significantly worse performance. This pattern is consistent and does not happen on Windows.

I'm not complaining anything. Just curious if any other users (especially stylo) has ever encountered this platform problem. Or if this is expected. The bahavior is so bad on Linux that I can't draw a conclusion whether the parallelism helps or not.

**Comment by adamreichold** [comment]
> Unlike on Windows, the app immediately starts sharp throttling after ~5 seconds. The more thread I open up, the harsher the throttling kicks in.

If I understand you correctly, you mean thermal throttling here? If so, could you check which cpufreq driver you are using on your Gentoo system? For example, by running `cpupower frequency-info`. I think for Intel CPU, this could be `acpi_cpufreq` and `intel_pstate` and trying the other one might yield better behaviour, but admittedly this is just a guess. See https://wiki.archlinux.org/title/CPU_frequency_scaling#Scaling_drivers for further information.
